handler/src/main/java/io/netty/handler/ssl/OpenSslKeyMaterialManager.java

```diff
@@ -0,0 +1,121 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import org.apache.tomcat.jni.SSL;
+
+import javax.net.ssl.SSLException;
+import javax.net.ssl.X509KeyManager;
+import javax.security.auth.x500.X500Principal;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Manages key material for {@link OpenSslEngine}s and so set the right {@link java.security.PrivateKey}s and
+ * {@link javax.security.cert.X509Certificate}s.
+ */
+class OpenSslKeyMaterialManager {
+
+    // Code in this class is inspired by code of conscrypts:
+    // - https://android.googlesource.com/platform/external/
+    //   conscrypt/+/master/src/main/java/org/conscrypt/OpenSSLEngineImpl.java
+    // - https://android.googlesource.com/platform/external/
+    //   conscrypt/+/master/src/main/java/org/conscrypt/SSLParametersImpl.java
+    //
+    static final String KEY_TYPE_RSA = "RSA";
+    static final String KEY_TYPE_DH_RSA = "DH_RSA";
+    static final String KEY_TYPE_EC = "EC";
+    static final String KEY_TYPE_EC_EC = "EC_EC";
+    static final String KEY_TYPE_EC_RSA = "EC_RSA";
+
+    // key type mappings for types.
+    private static final Map<String, String> KEY_TYPES = new HashMap<String, String>();
+    static {
+        KEY_TYPES.put("RSA", KEY_TYPE_RSA);
+        KEY_TYPES.put("DHE_RSA", KEY_TYPE_RSA);
+        KEY_TYPES.put("ECDHE_RSA", KEY_TYPE_RSA);
+        KEY_TYPES.put("ECDHE_ECDSA", KEY_TYPE_EC);
+        KEY_TYPES.put("ECDH_RSA", KEY_TYPE_EC_RSA);
+        KEY_TYPES.put("ECDH_ECDSA", KEY_TYPE_EC_EC);
+        KEY_TYPES.put("DH_RSA", KEY_TYPE_DH_RSA);
+    }
+
+    private final X509KeyManager keyManager;
+    private final String password;
+
+    OpenSslKeyMaterialManager(X509KeyManager keyManager, String password) {
+        this.keyManager = keyManager;
+        this.password = password;
+    }
+
+    void setKeyMaterial(OpenSslEngine engine) throws SSLException {
+        long ssl = engine.sslPointer();
+        String[] authMethods = SSL.authenticationMethods(ssl);
+        for (String authMethod : authMethods) {
+            String type = KEY_TYPES.get(authMethod);
+            if (type != null) {
+                setKeyMaterial(ssl, chooseServerAlias(engine, type));
+            }
+        }
+    }
+
+    void setKeyMaterial(OpenSslEngine engine, String[] keyTypes, X500Principal[] issuer) throws SSLException {
+        setKeyMaterial(engine.sslPointer(), chooseClientAlias(engine, keyTypes, issuer));
+    }
+
+    private void setKeyMaterial(long ssl, String alias) throws SSLException {
+        long keyBio = 0;
+        long keyCertChainBio = 0;
+        try {
+            // TODO: Should we cache these and so not need to do a memory copy all the time ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)